src/waveform/renderers/qopengl/waveformrendererendoftrack.cpp

```diff
@@ -0,0 +1,133 @@
+#include "waveform/renderers/qopengl/waveformrendererendoftrack.h"
+
+#include <QDomNode>
+#include <QPaintEvent>
+#include <QPainter>
+#include <memory>
+
+#include "control/controlobject.h"
+#include "control/controlproxy.h"
+#include "util/timer.h"
+#include "waveform/waveformwidgetfactory.h"
+#include "waveform/widgets/qopengl/waveformwidget.h"
+#include "widget/wskincolor.h"
+#include "widget/wwidget.h"
+
+namespace {
+
+constexpr int kBlinkingPeriodMillis = 1000;
+
+} // anonymous namespace
+
+using namespace qopengl;
+
+WaveformRendererEndOfTrack::WaveformRendererEndOfTrack(
+        WaveformWidgetRenderer* waveformWidget)
+        : WaveformShaderRenderer(waveformWidget),
+          m_pEndOfTrackControl(nullptr),
+          m_pTimeRemainingControl(nullptr) {
+}
+
+WaveformRendererEndOfTrack::~WaveformRendererEndOfTrack() {
+}
+
+bool WaveformRendererEndOfTrack::init() {
+    m_timer.restart();
+
+    m_pEndOfTrackControl.reset(new ControlProxy(
+            m_waveformRenderer->getGroup(), "end_of_track"));
+    m_pTimeRemainingControl.reset(new ControlProxy(
+            m_waveformRenderer->getGroup(), "time_remaining"));
+
+    return true;
+}
+
+void WaveformRendererEndOfTrack::setup(const QDomNode& node, const SkinContext& context) {
+    m_color = QColor(200, 25, 20);
+    const QString endOfTrackColorName = context.selectString(node, "EndOfTrackColor");
+    if (!endOfTrackColorName.isNull()) {
+        m_color.setNamedColor(endOfTrackColorName);
+        m_color = WSkinColor::getCorrectColor(m_color);
+    }
+    // m_pen = QPen(QBrush(m_color), 2.5 * scaleFactor());
+}
+
+void WaveformRendererEndOfTrack::initializeGL() {
+    QString vertexShaderCode = QStringLiteral(R"--(
+attribute vec4 position;
+attribute float gradient;
+varying float vgradient;
+void main()
+{
+    vgradient = gradient;
+    gl_Position = position;
+}
+)--");
+
+    QString fragmentShaderCode = QStringLiteral(R"--(
+uniform vec4 color;
+varying float vgradient;
+void main()
+{
+    gl_FragColor = vec4(color.x, color.y, color.z, color.w * (0.5 + 0.33 * max(0.,vgradient)));
+}
+)--");
+
+    initShaders(vertexShaderCode, fragmentShaderCode);
+}
+
+void WaveformRendererEndOfTrack::fillWithGradient(QColor color) {
+    const float positionArray[] = {-1.f, -1.f, 1.f, -1.f, -1.f, 1.f, 1.f, 1.f};
+    const float verticalGradientArray[] = {1.f, 1.f, -1.f, -1.f};
+    const float horizontalGradientArray[] = {-1.f, 1.f, -1.f, 1.f};
```

Comment:
I think these should be `static constexpr` globals in its own anonymous namespace.

Reply:
Agreed. Done.